### PR TITLE
Migrate snowflake crawler to using ACCOUNT [sc-10159]

### PR DESCRIPTION
**Pull request by alyiwang**
### 🤔 Why?

Most of the snowflake table information can be fetched from ACCOUNT_USAGE, instead of each database's INFORMATION_SCHEMA. This simplifies the metadata fetching process, and requires less permission to run.

### 🤓 What?

- fetch table and column info from ACCOUNT_USAGE
- update README with less permission requirement

### 🧪 Tested?

tested locally against Snowflake instance.

### CODE REVIEW

File: metaphor/snowflake/extractor.py [hunk 1]
```diff
@@ -539,11 +537,12 @@ def _init_dataset(
...
-        dataset.statistics = DatasetStatistics()
-        if row_count:
-            dataset.statistics.record_count = float(row_count)
-        if table_bytes:
-            dataset.statistics.data_size = table_bytes / (1000 * 1000)  # in MB
+        dataset.statistics = DatasetStatistics(
+            record_count=float(row_count) if row_count is not None else None,
+            data_size=table_bytes / (1000 * 1000)
+            if table_bytes is not None
+            else None,  # in MB
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:544]: The comment `# in MB` should be in line:542 ?

**Review by elic-eon — COMMENTED** [review]
Otherwise, LGTM

File: metaphor/snowflake/README.md [hunk 2]
```diff
@@ -19,18 +19,10 @@ set password = '<password>';
...
-grant references on future tables in database identifier($db) to role identifier($role);
-grant references on all views in database identifier($db) to role identifier($role);
-grant references on future views in database identifier($db) to role identifier($role);
-grant references on all materialized views in database identifier($db) to role identifier($role);
-grant references on future materialized views in database identifier($db) to role identifier($role);
 -- Grant permissions to access the snowflake "Account Usage" views:
 grant imported privileges on database snowflake to role identifier($role);
 -- If there are inbound shared databases, grant permissions to "show shares"
-grant imported share on account to identifier($role);
+grant import share on account to identifier($role);
```
[the remark this answers is not in the document]
Comment (alyiwang) [reply metaphor/snowflake/README.md:25]: Yeah, it was a typo before, corrected here.

File: metaphor/snowflake/extractor.py (hunk 1 above)
Comment (alyiwang) [reply metaphor/snowflake/extractor.py]: Thanks for catching, it's auto code format. Updated.

**Review by elic-eon — DISMISSED** [review]

File: metaphor/snowflake/README.md [hunk 3]
```diff
@@ -49,6 +41,21 @@ If you intend to use key pair authentication instead of password, follow the [Sn
 alter user identifier($user) set rsa_public_key='<public_key_content>';
 ```
 
+### Fetching extra info (Optional)
+
+If you intend to fetch extra table information such as LAST_CHANGE_COMMIT_TIME or DDL (enabled by default), following additional grant needs to be given to the newly created role.
+
+```sql
+grant usage on all schemas in database identifier($db) to role identifier($role);
```
Comment (mars-lan) [inline metaphor/snowflake/README.md:49]: Were you able to verify the permission in the end?

File: metaphor/snowflake/README.md [hunk 4]
```diff
@@ -49,6 +41,21 @@ If you intend to use key pair authentication instead of password, follow the [Sn
 alter user identifier($user) set rsa_public_key='<public_key_content>';
 ```
 
+### Fetching extra info (Optional)
+
+If you intend to fetch extra table information such as LAST_CHANGE_COMMIT_TIME or DDL (enabled by default), following additional grant needs to be given to the newly created role.
```
Comment (mars-lan) [inline metaphor/snowflake/README.md:46]:
```suggestion
To fetch extra table information such as the last modified time or DDL, please grant the following additional privileges to the newly created role.
```

File: metaphor/snowflake/README.md [hunk 5]
```diff
@@ -91,6 +98,14 @@ See [Output Config](../common/docs/output.md) for more information on `output`.
 
 See [Filter Configurations](../common/docs/filter.md) for more information on the optional `filter` config.
 
+#### Fetch Extra Table Information
+
+By default, the crawler fetches extra table information such as LAST_CHANGE_COMMIT_TIME or DDL. This requires extra grant to all the tables. To disable this, add the following:
```
Comment (mars-lan) [inline metaphor/snowflake/README.md:103]:
```suggestion
By default, the connector fetches extra table information such as LAST_CHANGE_COMMIT_TIME or DDL. This requires additional grants to all the tables. The connector will print warning messages if it has insufficient permissions. To disable this feature altogether, add the following to your config:
```

File: metaphor/snowflake/README.md (hunk 3 above)
Comment (alyiwang) [reply metaphor/snowflake/README.md:50]: Verified the permission grants, both for basic info (don't need table grant) or extra info (require table grants).

**Comment by mars-lan** [comment]
Need to bump up the coverage to pass the CI.

**Comment by alyiwang** [comment]
> Need to bump up the coverage to pass the CI.

Working on it. Is there a place we can check the lines/branches that are missing coverage?

**Comment by mars-lan** [comment]
> > Need to bump up the coverage to pass the CI.
> 
> 
> 
> Working on it. Is there a place we can check the lines/branches that are missing coverage?

Yes check the coveralls link in CI checks.

File: metaphor/snowflake/extractor.py [hunk 6]
```diff
@@ -91,28 +92,19 @@ async def extract(self) -> Collection[ENTITY_TYPES]:
...
-                    logger.info(f"Skip empty database {database}")
-                    continue
+            tables = self._fetch_tables(cursor, databases)
+            self._fetch_columns(cursor, databases)
+            self._fetch_primary_keys(cursor)
+            self._fetch_unique_keys(cursor)
 
-                logger.info(f"Include {len(tables)} tables from {database}")
+            if self._fetch_extra_info:
+                shared_databases = self._fetch_shared_databases(cursor)
```
Comment (mars-lan) [inline metaphor/snowflake/extractor.py:104]: Let's be defensive and do a try-catch here, in case the user misconfigured the permission?

**Review by mars-lan — DISMISSED** [review]

File: metaphor/snowflake/extractor.py (hunk 6 above)
Comment (alyiwang) [reply metaphor/snowflake/extractor.py]: updated

**Review by mars-lan — APPROVED** [review]

**Pull request merged**